### PR TITLE
Further testing updates & login validation

### PR DESCRIPTION
**Pull request by Oaphi**
The PR fixes security vulnerabilities in @babel/cli package, adds an email validation package to the `login` method of the `Client` class, and optimizes + adds new tests for several modules. Additionally, `IProfileData` is now exported to help other packages integrate with type safety, 

+ `Message` module now has `MessageAttributes` interface exported and used in the constructor to properly type the `attrs` parameter

### CODE REVIEW

**Comment by Oaphi** [comment]
very small changes are described in full in commit messages

**Review by samliew — APPROVED** [review]
Merge when ready

File: src/Browser.ts
Comment (danbopes) [inline src/Browser.ts:344]: Why are we initializing content here? If the server changes the message (Parses it for HTML, or other things), that will not be reflected on content anymore apart of message.

File: src/Client.ts
Comment (danbopes) [inline src/Client.ts:117]: Do we really need to do email validation here? I think we should simply let the server return the error in the event they use an invalid email to reduce the amount of dependencies down.

Comment (Oaphi) [reply src/Client.ts:117]: Well, if we are already doing input validation, why not validate the email without making an actual request? This way we know exactly what went wrong and can tell that to the user as early as possible. The package used is also well-known and lightweight - it's not like there are too many dependencies used as well.

Comment (danbopes) [reply src/Client.ts:117]: I was just basically doing very simple validation, make sure it exists/isnt undefined in the case of a variable/config issue or what not. I'd say this is fine, but just seems a little excessive.

File: src/Browser.ts
Comment (Oaphi) [reply src/Browser.ts:344]: Good point. It caused friction with the unit test but I guess I can mock the whole message module and thus mock the `_scrapeTranscript` private method to return the `text`. I am also not sure whether it should not be considered a failure state if the server returns a different message than the one sent. Is there a valid use case where we need to know the server changed something? Come to think of it - maybe should throw or have a flag on the message that identifies if the text passed differs from the text received?

File: src/Client.ts
Comment (Oaphi) [reply src/Client.ts:117]: hm... My reasoning was that the server could err in unpredicted ways, right? It is not a guarantee (ok, knowing how SE treats chat it can be a safe bet nothing changes) that the request fails or fails in the same way, thus we cannot reliably tell the user what went wrong, only that their message failed. And here we save one network round trip + tell the user exactly what's wrong for the price of one small package. Although I think the check can be trimmed further because I still left the `!email` check but, actually, I think it is not needed - as validation will fail on an empty email.

File: src/Browser.ts
Comment (Oaphi) [reply src/Browser.ts:344]: To be honest, I do not know what to do about that one - if you think that's important, I'll revert that one and mock out the module for the unit test to work.

Comment (Oaphi) [reply src/Browser.ts]: @danbopes - done, finally got time to rework the test to mock out the `Message` module instead and removed the direct passing of the content to the message instance. Let me know if the PR is ready to merge - I'd still like to get your seal of approval first before proceeding.

**Comment by danbopes** [comment]
@Oaphi I started doing a review, and as I was going through, I found a bunch of stuff I wanted to improve upon, so I ended up rewriting some major chunks. Can you review the development branch, check out the changes I made, and review my work?

I opted to move towards true private setters on the User class instead of the WeakMap, as # private accessor is now supported in babel. There was some massive cleanup, more documentation, some additional coverage (Although my latest commit will have added things that could use some more coverage).

**Comment by Oaphi** [comment]
> @Oaphi I started doing a review, and as I was going through, I found a bunch of stuff I wanted to improve upon, so I ended up rewriting some major chunks. Can you review the development branch, check out the changes I made, and review my work?
> 
> I opted to move towards true private setters on the User class instead of the WeakMap, as # private accessor is now supported in babel. There was some massive cleanup, more documentation, some additional coverage (Although my latest commit will have added things that could use some more coverage).

oh, I think we clashed a bit because I noticed your changes and started updating the `Message` module to accomodate the switch to the private fields because they have one drawback that originally prompted me to use a `WeakMap` instead: they cannot be iterated upon using dynamic propery names (see `_scrapeTranscript` method - but maybe you fixed that already). I have a bunch of merge conflicts so I'll need to take a look into that first

**Comment by danbopes** [comment]
> > @Oaphi I started doing a review, and as I was going through, I found a bunch of stuff I wanted to improve upon, so I ended up rewriting some major chunks. Can you review the development branch, check out the changes I made, and review my work?
> > I opted to move towards true private setters on the User class instead of the WeakMap, as # private accessor is now supported in babel. There was some massive cleanup, more documentation, some additional coverage (Although my latest commit will have added things that could use some more coverage).
> 
> oh, I think we clashed a bit because I noticed your changes and started updating the `Message` module to accommodate the switch to the private fields because they have one drawback that originally prompted me to use a `WeakMap` instead: they cannot be iterated upon using dynamic property names (see `_scrapeTranscript` method - but maybe you fixed that already). I have a bunch of merge conflicts so I'll need to take a look into that first.
> 
> Re: first glance - I see, you now save the whole result of the transcript into a private variable which solves the problem. I'll wade through to see what I can salvage on my side then take a look at the updates, sounds good?

I did notice that with the private variables, so that's why I simply used one variable for the whole object.

**Comment by danbopes** [comment]
We may want to hold off on the reconnecting websocket. It's nice and all, but doesn't exactly support if a user gets banned from the room. It will continually try to reconnect I'm guessing, which would be bad.

**Comment by Oaphi** [comment]
> We may want to hold off on the reconnecting websocket. It's nice and all, but doesn't exactly support if a user gets banned from the room. It will continually try to reconnect I'm guessing, which would be bad.

hm, can we unambiguously determine if the user is kicked/banned from a room? Or maybe keep a counter of disconnects and set a relatively low threshold assuming that, say, after X disconnects we will not attempt to reconnect - what do you think?

**Comment by danbopes** [comment]
> > We may want to hold off on the reconnecting websocket. It's nice and all, but doesn't exactly support if a user gets banned from the room. It will continually try to reconnect I'm guessing, which would be bad.
> 
> hm, can we unambiguously determine if the user is kicked/banned from a room? Or maybe keep a counter of disconnects and set a relatively low threshold assuming that, say, after X disconnects we will not attempt to reconnect - what do you think?

That's entirely up to you at this point. I didn't want to test/try to figure out all nuisances with how the websocket ultimately gets disconnected server side. I just assume: If I'm not the one initiating the disconnect, then I should simply try to reconnect. Maybe a simple pushback system, that retries immediately and then adds little short delays maybe up to 2 minute/5 minute intervals.

**Comment by Oaphi** [comment]
> > > We may want to hold off on the reconnecting websocket. It's nice and all, but doesn't exactly support if a user gets banned from the room. It will continually try to reconnect I'm guessing, which would be bad.
> > 
> > 
> > hm, can we unambiguously determine if the user is kicked/banned from a room? Or maybe keep a counter of disconnects and set a relatively low threshold assuming that, say, after X disconnects we will not attempt to reconnect - what do you think?
> 
> That's entirely up to you at this point. I didn't want to test/try to figure out all nuisances with how the websocket ultimately gets disconnected server side. I just assume: If I'm not the one initiating the disconnect, then I should simply try to reconnect. Maybe a simple pushback system, that retries immediately and then adds little short delays maybe up to 2 minute/5 minute intervals.

Yeah, actually, back-off sounds like a nice solution, thank you - I'll try to find a way to determine the room ban and then default to it if I come up with nothing (which is pretty likely).

**Review by samliew — APPROVED** [review]

**Pull request merged**
